src/client.rs

```diff
@@ -161,14 +162,175 @@ impl Client {
+        self.count(metric, 1.0);
+    }
+
+    /// Decrement a metric by -1
+    ///
+    /// ```ignore
+    /// # Decrement a given metric by 1
+    /// pipe = client.pipeline();
+    /// pipe.decr("metric.completed");
+    /// pipe.send();
+    /// ```
+    ///
+    /// This modifies a counter with an effective sampling
+    /// rate of 1.0.
+    pub fn decr(&mut self, metric: &str) {
+        self.count(metric, -1.0);
+    }
+
+    /// Modify a counter by `value`.
+    ///
+    /// Will increment or decrement a counter by `value` with
+    /// a sampling rate of 1.0.
+    ///
+    /// ```ignore
+    /// # Increment by 12
+    /// pipe = client.pipeline();
+    /// pipe.count("metric.completed", 12);
+    /// pipe.send();
+    /// ```
+    pub fn count(&mut self, metric: &str, value: f64) {
+        let data = self.client.prepare(format!("{}:{}|c", metric, value));
+        self.stats.push_back(data);
+    }
+
+    /// Modify a counter by `value` only x% of the time.
+    ///
+    /// Will increment or decrement a counter by `value` with
+    /// a custom sampling rate.
+    ///
+    ///
+    /// ```ignore
+    /// # Increment by 4 50% of the time.
+    /// pipe = client.pipeline();
+    /// pipe.sampled_count("metric.completed", 4, 0.5);
+    /// pipe.send();
+    /// ```
+    pub fn sampled_count(&mut self, metric: &str, value: f64, rate: f64) {
+        if rand::random::<f64>() < rate {
+            return;
+        }
+        let data = self.client.prepare(format!("{}:{}|c", metric, value));
+        self.stats.push_back(data);
+    }
+
+    /// Set a gauge value.
+    ///
+    /// ```ignore
+    /// # set a gauge to 9001
+    /// pipe = client.pipeline();
+    /// pipe.gauge("power_level.observed", 9001);
+    /// pipe.send();
+    /// ```
+    pub fn gauge(&mut self, metric: &str, value: f64) {
+        let data = self.client.prepare(format!("{}:{}|g", metric, value));
+        self.stats.push_back(data);
+    }
+
+    /// Send a timer value.
+    ///
+    /// The value is expected to be in ms.
+    ///
+    /// ```ignore
+    /// # pass a duration value
+    /// pipe = client.pipeline();
+    /// pipe.timer("response.duration", 10.123);
+    /// pipe.send();
+    /// ```
+    pub fn timer(&mut self, metric: &str, value: f64) {
+        let data = self.client.prepare(format!("{}:{}|ms", metric, value));
+        self.stats.push_back(data);
+    }
+
+    /// Time a block of code.
+    ///
+    /// The passed closure will be timed and executed. The block's
+    /// duration will be sent as a metric.
+    ///
+    /// ```ignore
+    /// # pass a duration value
+    /// pipe = client.pipeline();
+    /// pipe.time("response.duration", || {
+    ///   # Your code here.
+    /// });
+    /// pipe.send();
+    /// ```
+    pub fn time<F>(&mut self, metric: &str, callable: F)
+        where F: Fn()
+    {
+        let start = clock_ticks::precise_time_ms();
+        callable();
+        let end = clock_ticks::precise_time_ms();
```

Comment:
I wonder if there is a way we could avoid duplicating this and some of the other method logic. Perhaps if the `Client` was wrapped in a decorator that provided a stubbed `send()` method? That might let `Pipeline` simply delegate to the pipeline 'client'.